libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+				connect(m_currentNode, nextNode);
+				m_currentNode = nextNode;
+				return false;
+			}
+			default:
+				break;
+		}
+	return ASTConstVisitor::visit(_functionCall);
+}
+
+bool CFG::constructFlow(ASTNode const& _astRoot)
+{
+	_astRoot.accept(*this);
+	applyModifiers();
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
+}
+
+
+bool CFG::visit(ModifierDefinition const& _modifier)
+{
+	m_currentModifierFlow = make_shared<ModifierFlow>(newNode(), newNode(), newNode());
+	m_currentModifierFlow->placeholderEntry = newNode();
+	m_currentModifierFlow->placeholderExit = newNode();
+	m_currentFunctionFlow = std::static_pointer_cast<FunctionFlow>(m_currentModifierFlow);
+	m_modifierControlFlow[&_modifier] = m_currentModifierFlow;
+
+	ControlFlowParser::createFlowFromTo(*this, m_currentFunctionFlow->entry, m_currentFunctionFlow->exit, _modifier);
+
+	m_currentModifierFlow.reset();
+	m_currentFunctionFlow.reset();
+
+	return false;
+}
+
+bool CFG::visit(FunctionDefinition const& _function)
+{
+	m_currentFunctionFlow = make_shared<FunctionFlow>(newNode(), newNode(), newNode());
+	m_functionControlFlow[&_function] = m_currentFunctionFlow;
+
+	ControlFlowParser::createFlowFromTo(*this, m_currentFunctionFlow->entry, m_currentFunctionFlow->exit, _function);
+
+	m_currentFunctionFlow.reset();
+
+	return false;
+}
+
+FunctionFlow const& CFG::functionFlow(FunctionDefinition const& _function) const
+{
+	solAssert(m_functionControlFlow.count(&_function), "");
+	return *m_functionControlFlow.find(&_function)->second;
+}
+
+CFGNode* CFG::newNode()
+{
+	m_nodes.emplace_back(new CFGNode());
+	return m_nodes.back().get();
+}
+
+void CFG::applyModifiers()
+{
+	for (auto const& function: m_functionControlFlow)
+	{
+		for (auto const& modifierInvocation: boost::adaptors::reverse(function.first->modifiers()))
+		{
+			if (auto modifierDefinition = dynamic_cast<ModifierDefinition const*>(
+				modifierInvocation->name()->annotation().referencedDeclaration
+			))
+			{
+				solAssert(m_modifierControlFlow.count(modifierDefinition), "");
+				auto modifierFlow = m_modifierControlFlow[modifierDefinition];
+				applyModifierFlowToFunctionFlow(*modifierFlow, function.second);
+			}
+		}
+	}
+}
+
+void CFG::applyModifierFlowToFunctionFlow(
+	ModifierFlow const& _modifierFlow,
+	std::shared_ptr<FunctionFlow> _functionFlow
+)
+{
+	map<CFGNode*, CFGNode*> oldToNew;
+
+	// inherit the revert node of the function
+	oldToNew[_modifierFlow.revert] = _functionFlow->revert;
+
+	// replace the placeholder nodes by the function entry and exit
+	oldToNew[_modifierFlow.placeholderEntry] = _functionFlow->entry;
+	oldToNew[_modifierFlow.placeholderExit] = _functionFlow->exit;
+
+	stack<CFGNode*> nodesToClone;
+	nodesToClone.push(_modifierFlow.entry);
+
+	// map the modifier entry to a new node that will become the new function entry
+	oldToNew[_modifierFlow.entry] = newNode();
+
+	while (!nodesToClone.empty())
+	{
+		auto srcNode = nodesToClone.top();
```

Comment:
I'm sorry, but I am really not fond of `auto`, I constantly have to look up the type to check whether it performs a copy or not.